lisa/tests/staging/schedutil.py

```diff
@@ -0,0 +1,271 @@
+
+    def _plot_test_boost(self, df):
+        task = self.rtapp_tasks[0]
+        analysis = self.trace.analysis.frequency
+        fig, axis = analysis.setup_plot(interactive=False)
+        df['cost_margin'].plot(ax=axis, drawstyle='steps-post', color='r')
+        df['boost_points'].astype('int', copy=False).plot(ax=axis, drawstyle='steps-post', color='black')
+        df['expected_cost_margin'].plot(ax=axis, drawstyle='steps-post', color='blue')
+
+        self.trace.analysis.tasks.plot_task_activation(task, axis=axis, overlay=True)
+
+        axis.legend()
+        analysis.save_plot(fig, filepath=os.path.join(self.res_dir, 'ramp_boost.svg'))
+        return axis
+
+    @requires_events('sched_switch')
+    def trace_window(self, trace):
+        """
+        Skip the first phase in the trace, since it is heavily influenced by
+        what was immediately preceding it.
+        """
+        profile = self.rtapp_profile
+        phase_duration = max(
+            task.phases[0].duration_s
+            for task in profile.values()
+        )
+        start, end = super().trace_window(trace)
+        return (start + phase_duration, end)
+
+    @df_ramp_boost.used_events
+    def test_ramp_boost(self, nrg_threshold_pct=0.1, bad_samples_threshold_pct=0.1) -> ResultBundle:
+        """
+        Test that the energy boost feature is triggering as expected.
+        """
+        # If there was no cost_margin sample to look at, that means boosting
+        # was not exhibited by that test so we cannot conclude anything
+        df = self.df_ramp_boost()
+        self._plot_test_boost(df)
+
+        if df.empty:
+            return ResultBundle(Result.UNDECIDED)
+
+        # Make sure the boost is always positive (negative cannot really happen
+        # since the kernel is using unsigned arithmetic, but still check in
+        # case there are some dataframe handling issues)
+        assert not (df['expected_cost_margin'] < 0).any()
+        assert not (df['cost_margin'] < 0).any()
+
+        # "rect" method is accurate here since the signal is really following
+        # "post" steps
+        expected_boost_nrg = series_mean(df['expected_cost_margin'])
+        actual_boost_nrg = series_mean(df['cost_margin'])
+
+        # Check that the total amount of boost is close to expectations
+        lower = max(0, expected_boost_nrg - nrg_threshold_pct)
+        higher = expected_boost_nrg
+        passed_overhead = lower <= actual_boost_nrg <= higher
+
+        # Check the shape of the signal: actual boost must be lower or equal
+        # than the expected one.
+        good_shape_nr = (df['cost_margin'] <= df['expected_cost_margin']).sum()
+
+        df_len = len(df)
+        bad_shape_nr = df_len - good_shape_nr
+        bad_shape_pct = bad_shape_nr / df_len * 100
+
+        # Tolerate a few bad samples that added too much boost
+        passed_shape = bad_shape_pct < bad_samples_threshold_pct
+
+        passed = passed_overhead and passed_shape
```

Comment:
Would it make sense to split these into two separate tests? We'd get finer grained results.

Reply:
I'm not sure if that would really help, all these checks are just checking the overall shape of the signal. If I could simulate the signals with 100% accuracy, they would all go away and replaced by a Series equality check.